persistence/relational-jdbc/src/main/java/org/apache/polaris/persistence/relational/jdbc/JdbcBasePersistenceImpl.java

```diff
@@ -431,15 +448,34 @@ public PolarisBaseEntity lookupEntityByName(
             name,
             "realm_id",
             realmId);
-    return getPolarisBaseEntity(
+    QueryGenerator.PreparedQuery query =
         QueryGenerator.generateSelectQuery(
-            ModelEntity.getAllColumnNames(schemaVersion), ModelEntity.TABLE_NAME, params));
+            ModelEntity.getAllColumnNames(schemaVersion), ModelEntity.TABLE_NAME, params);
+    return getPolarisBaseEntity(query, connection);
   }
 
   @Nullable
   private PolarisBaseEntity getPolarisBaseEntity(QueryGenerator.PreparedQuery query) {
+    return getPolarisBaseEntity(query, null);
+  }
+
+  /**
+   * Execute entity lookup, optionally using an existing connection to maintain transaction
+   * consistency.
+   *
+   * @param query the prepared query to execute
+   * @param connection optional connection to reuse; if null, a new connection will be obtained
+   * @return the entity if found, null otherwise
+   */
+  @Nullable
+  private PolarisBaseEntity getPolarisBaseEntity(
+      QueryGenerator.PreparedQuery query, @Nullable Connection connection) {
     try {
-      var results = datasourceOperations.executeSelect(query, new ModelEntity(schemaVersion));
+      List<PolarisBaseEntity> results =
+          connection != null
```

Comment:
It looks like `connection` is `null` when called from `writeEntity`... Does that call path not have the same issue (reading in another transaction) that this PR is trying to fix?

Reply:
I mean that `writeEntity` invokes this select as part of its execution, but the select will run in a new JDBC transaction (because `connection` is `null`). Basically, the change `persistEntity` (called by ` writeEntity`) will make and the select it may do on exception will run in different connections in runtime. The code of `persistEntity` declares a connection parameter, which makes the impression that _all_ the logic in that method runs over the same connection :thinking: 

I agree that there is no bug ATM, but the logic is very obscure now that connection may or may not be `null` (new).

Would it be possible to refactor the code such that `connection` parameters a non-null when present?

Reply:
Attempted this in commit: https://github.com/apache/polaris/pull/3334/changes/192d7870869216fe24ba50a1a7fb36005c698d48, wdyt ?